Wallet database handling abstractions/simplifications

laanwj opened the pull request:
My immediate goal with this is to make it easier to use a different key/value store instead of BerkeleyDB, for a sandboxing project. However, it is an improvement also for BerkeleyDB.

Steps:

- Abstract database handle from explicit strFilename  into CWalletDBWrapper. For many databases, a usable handle involves more than just a filename. Also make the field private.

- Get rid of fFileBacked - it is only used by the tests, and used inconsistly - for example CWallet::AddToWallet doesn't check it at all, and that's not quite the only place (see AbandonWallet). Fixing would be a lot of  work and lead to disjointed code.

  But it is not necessary. Instead move the concern to ignoring operations with a dummy database to CDB. There were very few changes necessary for this.

- Reduce references to global bitdb. CWalletDBWrapper carries this environment information around so there is less need to refer to global scope. Besides being cleaner, this could allow use of multiple database environments at some point.

- CWalletDB now contains a CDB instead of inheriting from it. This makes it easier to replace the internal transaction with a different database, without leaking through internals.

Future:

- The eventual goal of "CWalletDBWrapper" is like "CDBWrapper", to move further database functionality there from "CDB" (which is really a "CDBBatch", a database transaction object). Not renaming it nor CWalletDB here to not interfere unduly much with other open wallet pulls.

- Two maps inside bitdb that refer to wallet files by name could be removed,
  and converted to fields on CWalletDBWrapper:

    std::map<std::string, int> mapFileUseCount;
    std::map<std::string, Db*> mapDb;

laanwj committed:
wallet: Introduce database handle wrapper

Abstract database handle from explicit strFilename into
CWalletDBWrapper.

Also move CWallet::Backup to db.cpp - as it deals with representation
details this is a database specific operation.

laanwj committed:
wallet: Get rid of fFileBacked

Instead, CWalletDB() with a dummy handle will just give you a no-op
database in which writes always succeeds and reads always fail. CDB
already had functionality for this, so just use that.

laanwj committed:
wallet: Reduce references to global bitdb environment

laanwj committed:
wallet: CWalletDB CDB composition not inheritance

CWalletDB now contains a CDB instead of inheriting from it.

This makes it easier to replace the internal transaction with a different
database, without leaking through internals.

jonasschnelli commented:
Great!
Concept ACK.
Needs rebase.

laanwj commented:
Yep needed trivial rebase for #9643.

laanwj commented on src/wallet/walletdb.h:
```diff
@@ -112,11 +112,19 @@ class CKeyMetadata
 {
 public:
-    CWalletDB(const std::string& strFilename, const char* pszMode = "r+", bool _fFlushOnClose = true) : CDB(strFilename, pszMode, _fFlushOnClose)
+    /** Takes a std::unique_ptr reference to avoid having to use
```
This unique_ptr here bothers me. The interface shouldn't care what kind of pointer is used. I think I'm going to change this to take a `CWalletDBWrapper&`, then replace the call sites with `CWalletDB(*dbw)`. That's better than `CWalletDB(dbw.get())` at least.

benma commented on src/wallet/db.cpp:
```diff
@@ -359,13 +359,12 @@ void CDBEnv::CheckpointLSN(const std::string& strFile)
     fFlushOnClose = fFlushOnCloseIn;
-    if (strFilename.empty())
-        return;
+    std::string strFilename = dbw.strWalletFile;
```
Better to use `const std::string& strFileName`, no copying and it's clear it won't be modified.

laanwj replied:
Agreed.

benma commented on src/wallet/db.cpp:
```diff
@@ -472,8 +471,11 @@ bool CDBEnv::RemoveDb(const std::string& strFile)
 {
+    if (!dbw.env)
+        return true;
+    std::string strFile = dbw.strWalletFile;
```
same

benma commented on src/wallet/db.cpp:
```diff
@@ -596,9 +598,13 @@ void CDBEnv::Flush(bool fShutdown)
+    if (!dbw) {
+        return true;
+    }
+    std::string strFile = dbw->strWalletFile;
```
same

benma commented on src/wallet/db.cpp:
```diff
@@ -596,9 +598,13 @@ void CDBEnv::Flush(bool fShutdown)
+bool CDB::PeriodicFlush(CWalletDBWrapper *dbw)
 {
     bool ret = false;
+    if (!dbw) {
```
Better to move the check to the top, above `bool ret`.

benma commented on src/wallet/db.cpp:
```diff
@@ -633,3 +639,44 @@ bool CDB::PeriodicFlush(std::string strFile)
+
+bool CWalletDBWrapper::Backup(const std::string& strDest)
+{
+    if (!env)
```
Inconsistent style with using/skipping braces (above, you used braces, here, you don't). I'd add them.

laanwj replied:
Yes the brace should be after

benma commented on src/wallet/db.h:
```diff
@@ -87,6 +87,33 @@ class CDBEnv
+    std::string GetName() { return strWalletFile; }
+
+    CDBEnv *env;
+    std::string strWalletFile;
```
const std::string

benma commented on src/wallet/db.h:
```diff
@@ -87,6 +87,33 @@ class CDBEnv
+
+    /** Get a name for this database, for debugging etc.
+     */
+    std::string GetName() { return strWalletFile; }
```
`std::string GetName() const`

You have this function, and the public `strWalletFile`, which you access directly in other places. Maybe `strWalletFile` should be made private?

laanwj replied:
They are subtly different, only happen to be the same for BerkeleyDB: 

- `GetName` is an an abstract name to refer to the database object for logging. It should exist for any version of CDBWrapper.
- `strWalletFile` is an implementation detail that only exists for BerkeleyDB. Other databases may have completely different connection descriptions, which do not involve simply a string. So this is private to the low-level implementation (CDB), the higher-level stuff (such as CWallet) should not refer to it.

laanwj replied:
Could make this clearer using "friend" I guess.

benma commented on src/wallet/wallet.h:
```diff
@@ -653,17 +653,35 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
+
+    /** Get a name for this wallet for logging/debugging purposes.
+     */
+    std::string GetName()
```
`GetName() const`

laanwj replied:
Yes good catch.

benma commented on src/wallet/wallet.h:
```diff
@@ -653,17 +653,35 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
+    {
+        if (dbw) {
+            return dbw->GetName();
+        } else {
```
Maybe it is a matter of personal preference, but I always find it odd to put an `else` when the `if` returns.

laanwj replied:
I prefer to keep it this way. We have no guideline on this in `doc/developer-notes.md`, and I personally find it easier to follow the control flow this way.

benma reviewed (commented):
Some comments on the first commit.

Since this is my first review, please let me know if it is overboard/too pedantic.

laanwj commented:
Fixed @benma's nits.

benma commented on src/wallet/db.h:
```diff
@@ -87,6 +87,36 @@ class CDBEnv
+private:
+    /** BerkeleyDB specific */
+    CDBEnv *env;
+    std::string strFile;
```
`const std::string strFile;` ?

benma commented:
For my understanding: the refactor to support a different db backend than BerkeleyDB is not finished with this PR, right? It is a bit hard to see through `CDBEnv`, `CDB` and `CWalletDBWrapper`, but `CDB` looks half-finished: it still refers to the walletFile string in most functions instead of going through `CWalletDBWrapper`.

Also, berkeley-specific fields are in `CWalletDBWrapper`. If you want to use a different backend that doesn't rely on a filename, where would that fit in? Would that end up in two classes inheriting from `CWalletDBWrapper`, and moving the berkeley-specific stuff out of it?

> The eventual goal of "CWalletDBWrapper" is like "CDBWrapper", to move further database functionality there from "CDB" (which is really a "CDBBatch", a database transaction object). Not renaming it nor CWalletDB here to not interfere unduly much with other open wallet pull

In the end, is there even a need for the class `CDB` anymore?

laanwj commented:
> For my understanding: the refactor to support a different db backend than BerkeleyDB is not finished with this PR, right? It is a bit hard to see through CDBEnv, CDB and CWalletDBWrapper, but CDB looks half-finished: it still refers to the walletFile string in most functions instead of going through CWalletDBWrapper.

To be clear: CDB and CWalletDBWrapper and CDBEnv are database specific. You need to replace those if you want to change the backend. In many cases you can do without CDBEnv (for LevelDB I could just delete it). So `db.h/cpp` is database specific, `walletdb.h/cpp` and `wallet.h/cpp` are not.

On top of this I have a patch that switches the backend to LevelDB, mostly changing only `db.cpp`/`db.h` (completely)  and CWalletDB constructor. I need this for testing, to be clear I'm not going to PR that for merging.

It's not quite as easy as switching the `dbwrapper.h` backend yet, but much easier than before.

laanwj commented:
@benma FYI not for review but if you're interested  https://github.com/laanwj/bitcoin/tree/20170310_experiment_leveldb_wallet is a leveldb wallet on top of (an old version of) this. It passes the unit and QA tests apart from `backupwallet` which isn't implemented.

TheBlueMatt commented on src/wallet/db.h:
```diff
@@ -118,7 +165,7 @@ class CDB
     CDB(const CDB&);
```
Can you go ahead and =delete these, since we have C++11?

laanwj replied:
Good idea

TheBlueMatt reviewed (commented):
Dear god this stuffs a mess, good thing you're cleaing it up :P.

utACK e3e51fdd3a9491ab4a07baed0a0f52b47ef70140

sipa commented:
Concept ACK

jonasschnelli commented:
Tested ACK e3e51fdd3a9491ab4a07baed0a0f52b47ef70140
(https://bitcoin.jonasschnelli.ch/build/PR/9951)

Though needs a "trivial" rebase.

JeremyRubin commented on src/wallet/db.cpp:
```diff
@@ -359,13 +359,12 @@ void CDBEnv::CheckpointLSN(const std::string& strFile)
 
 
-CDB::CDB(const std::string& strFilename, const char* pszMode, bool fFlushOnCloseIn) : pdb(NULL), activeTxn(NULL)
+CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb(NULL), activeTxn(NULL)
```
I'm a little confused on the API here.

I feel like the unwrapped version of CDB shouldn't be compatible with the wrapped version of CWalletDB?

laanwj replied:
CWalletDBWrapper is not a wrapper around CWalletDB. We kind of have a naming problem with the wallet:
- `CDBEnv` is an environment in which the database exists (has no analog in `dbwrapper.h`)
- `CWalletDBWrapper` represents a wallet database (similar to `CDBWrapper` in `dbwrapper.h`)
- `CDB` is a low-level database transaction (similar to `CDBBatch` in `dbwrapper.h`)
- `CWalletDB` is a modifier object for the wallet, and encapsulates a database transaction as well as methods to act on the database (no analog in `dbwrapper.h`)

The latter two are named badly, should be renamed at some point but didn't want to do it here.

laanwj replied:
I added a commit that adds this as a comment, as it will be something that confuses everyone reading this code.

JeremyRubin commented on src/wallet/db.h:
```diff
@@ -111,6 +117,12 @@ class CWalletDBWrapper
+     * Only to be used at a low level, application should ideally not care
+     * about this.
+     */
+    bool IsDummy() { return env == nullptr; }
```
If it's only to be used by low level stuff, maybe friend the function specifically to those classes...

laanwj replied:
Yes, good point. CDB is already a friend and should be the only user of this.

JeremyRubin commented on src/wallet/walletdb.cpp:
```diff
@@ -33,46 +33,46 @@ static std::atomic<unsigned int> nWalletDBUpdateCounter;
 {
     nWalletDBUpdateCounter++;
-    return Write(make_pair(std::string("name"), strAddress), strName);
+    return batch.Write(make_pair(std::string("name"), strAddress), strName);
```
missing std::

laanwj replied:
I don't understand why this doesn't cause a compile error. We've removed all the "using namespace std".

JeremyRubin reviewed (approved):
Couple nits, overall utack.

jonasschnelli commented:
Needs rebase (Imo this is ready for merge).

laanwj commented:
Yes, will rebase this soon(TM) and fix the nits.

laanwj committed:
wallet: Make IsDummy private in CWalletDBWrapper

This is only for use in the low-level functions, and CDB is already
a friend class.

laanwj committed:
wallet: Add comment describing the various classes in walletdb.h
